nx/lib/nx/defn.ex

```diff
@@ -267,6 +267,60 @@ defmodule Nx.Defn do
     define(:defp, call, block, __CALLER__)
   end
 
+  @doc """
+  Defines a public numerical function that delegates to another
+  module.
+
+  Delegated function signature patterns must match target defn
+  function signature patterns. For example, if the target function
+  has a 2-tuple as its first arg then the delegated function
+  signature must also have a 2-tuple as its first arg.
+
+  ## Examples
+
+      defmodule MyImpl do
+        import Nx.Defn
+
+        defn calc(a, b), do: a * b + 1
+        defn calc_pair({a, b}), do: calc(a, b)
+      end
+
+      defmodule MyMath do
+        import Nx.Defn
+
+        defndelegate calc(a, b), to: MyImpl
+
+        defndelegate calc_pair({a, b}), to: MyImpl
+
+        defndelegate calc_w_other_name(a, b), to: MyImpl, as: :calc
+
+        defndelegate calc_pair_tuple({a, b}), to: MyImpl, as: :calc_pair
+      end
+  """
+  defmacro defndelegate({name, meta, args}, opts) do
```

Comment:
Unfortunately, if we want to support all cases found in `Kernel`, which is what users would likely expect, this is slightly more complex. Perhaps it is best to copy and paste the implementation from Kernel and then replace `def` here by `defn`:

https://github.com/elixir-lang/elixir/blob/master/lib/elixir/lib/kernel.ex#L5282

Just to give one example of the differences, we can do this in Elixir:

    foo = MyMod
    defdelegate bar(arg1, arg2), to: foo

We can also delegate with default arguments and so on. The good news is that Elixir's implementation has some code that has been deprecated, so we can remove those deprecated APIs for `defn`.

Reply:
For some reason it didn't occur to me to try the `defdelegate` implementation. I'll swap implementations and try it out.